### PR TITLE
[Inductor] Fix OpenMP discovery on MacOS

### LINKED ISSUE #90362: TorchDynamo example crashes on Apple Silicon with `-march=native` flag not supported
### 🐛 Describe the bug

The torch dynamo code on Apple Silicon CPU is crashing. Minimal code example:

```python
import torch
from typing import List
import torch._dynamo

def toy(a, b):
    x = a / (torch.abs(a) + 1)
    if (b.sum() < 0):
        b = b* -1
    return x*b
compiled_toy = torch.compile(toy)
print(compiled_toy(torch.randn(4), torch.randn(4)))
```

The Output is:

```
BackendCompilerFailed: _compile_fn raised CppCompileError: C++ compile error

Command:
g++ /tmp/torchinductor_kulinseth/rf/crfks4ygnmnznafczrgu2spfbjzqasuvauzpfyqipjvsqvwpk2u5.cpp -shared -fPIC -Wall -std=c++14 -Wno-unused-variable -I/Users/kulinseth/miniconda3/envs/repro/lib/python3.9/site-packages/torch/include -I/Users/kulinseth/miniconda3/envs/repro/lib/python3.9/site-packages/torch/include/torch/csrc/api/include -I/Users/kulinseth/miniconda3/envs/repro/lib/python3.9/site-packages/torch/include/TH -I/Users/kulinseth/miniconda3/envs/repro/lib/python3.9/site-packages/torch/include/THC -I/Users/kulinseth/miniconda3/envs/repro/include/python3.9 -lgomp -march=native -O3 -ffast-math -fno-finite-math-only -fopenmp -D C10_USING_CUSTOM_GENERATED_MACROS -o/tmp/torchinductor_kulinseth/rf/crfks4ygnmnznafczrgu2spfbjzqasuvauzpfyqipjvsqvwpk2u5.so

Output:
clang: error: the clang compiler does not support '-march=native'
clang: error: unsupported option '-fopenmp'
clang: error: unsupported option '-fopenmp'


Set torch._dynamo.config.verbose=True for more information


You can suppress this exception and fall back to eager by setting:
    torch._dynamo.config.suppress_errors = True

```
Currently Clang doesn't have support for `-march=native` on Apple Silicon. This is currently on Monterey. I can check on latest Ventura if this flag is added.

### Versions

```
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: macOS 12.5 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.11.23)
CMake version: version 3.22.0
Libc version: N/A

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:01:00)  [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-12.5-arm64-arm-64bit
Is CUDA available: N/A
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

Versions of relevant libraries:
[pip3] numpy==1.23.2
[conda] numpy                     1.23.2           py39h3668e8b_0    conda-forge
```

cc @ezyang @gchanan @zou3519 @soumith @msaroufim @wconstab @ngimel @bdhirsh @mlazos @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### PR DESCRIPTION
**Pull request by malfet**
It's not available as system dependency, so assume that it is installed
using Anaconda

Also, clang on MacOS does not recognize `-fopenmp` flag, but according
to https://mac.r-project.org/openmp/ and local experiments `-Xclang
-fopenmp` always works

Test plan:
Following should run and return true
```python
import torch

def foo(x: torch.Tensor) -> torch.Tensor:
   return torch.sin(x) + torch.cos(x)

if __name__=="__main__":
    x = torch.rand(3, 3)
    x_eager = foo(x)
    x_pt2 = torch.compile(foo)(x)
    print(torch.allclose(x_eager, x_pt2))
```

Skip number of tests that fail on x86 MacOS (for example rsqrt for bool type and   `test_pixel_shuffle_channels_last_cpu` on machines that do not support AVX2)
Tweak few tests to use double precision when running on CPU, as type promotion for accumulator types is broken.

TODO: Fix PyTorch for M1 compilation with OpenMP, bundle `omp.h` into the package and use it instead.
Fixes https://github.com/pytorch/pytorch/issues/90362

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by jansel — APPROVED** [review]

**Review by jgong5 — APPROVED** [review]
LGTM.

**Comment by jgong5** [comment]
@EikanWang With this PR, do we still need this https://github.com/pytorch/pytorch/pull/90830?

File: torch/_inductor/codecache.py [hunk 1]
```diff
@@ -402,8 +409,24 @@ def get_include_and_linking_paths(
         # This approach allows us to only pay for what we use.
         ipaths = cpp_extension.include_paths() + [sysconfig.get_path("include")]
         lpaths = []
-        libs = ["gomp"]
         macros = ""
+        if sys.platform == "darwin":
+            # GNU OpenMP generally is not available on MacOS
+            # There is either Intel OpenMP(for x86) or LLVM OpenMP (for both x86 and arm64)
+            libs = ["omp"]
+            if os.getenv("CONDA_PREFIX") is not None:
```
Comment (albanD) [inline torch/_inductor/codecache.py:417]: I think we should give users more feedback related to this as I don't thing most users use conda.
In particular should we generally recommend people use conda if they want to use torch.compile? Should we do that on macos at least?
If they don't do that, then they are reponsible for setting up the right global deps.

If that is our stance, I think we should mention that in the doc + have a warning.

Comment (malfet) [reply torch/_inductor/codecache.py:417]: Our `torch.compile` CPU story is somewhat confusing:
- It plainly does not work on Windows
- Until this PR I don't know how one can make it work on MacOS (I could not find how to install gnu OpenMP on MacOS, not even sure if this is possible)
- It works on Linux only if openmp is installed 

But yeah, we probably should bundle `omp.h` into `torch/include` and use it from there by default, will probably do that in followup PR.

**Comment by malfet** [comment]
Ok, I've just learned a fun fact (and understand a bit more, why @voznesenskym want's to announce test bankruptcy):
Fail to compile means test will pass, but if it compiles, but yields wrong results -> failure

**Comment by malfet** [comment]
`CpuTests.test_pow2` is failing because of the mismatch between scalar type promotion for eager and inductor, see 
https://github.com/pytorch/pytorch/blob/37a28255cb9c2a78fd2a27ed7921e8c9672a57ab/aten/src/ATen/native/cpu/PowKernel.cpp#L112
vs
```
```

**Comment by malfet** [comment]
Found similar accuracy problem in `test_tmp_not_defined_issue2` where in eager reduction results are stored in double, but in `torch.compile` for CPU  in float.
And `test_pixel_shuffle_channels_last_cpu ` fails with fantastic `Your CPU does not support FBGEMM`

**Pull request closed without merging**